internal/controller/nonadminbackupstoragelocation_controller.go

```diff
@@ -0,0 +1,682 @@
+			Status:  metav1.ConditionTrue,
+			Reason:  "SecretUpdated",
+			Message: "Secret successfully updated in the OADP namespace",
+		})
+	case controllerutil.OperationResultNone:
+		logger.V(1).Info("VeleroBackupStorageLocation secret unchanged",
+			constant.NamespaceString, veleroBslSecret.Namespace,
+			constant.NameString, veleroBslSecret.Name)
+	}
+
+	if secretSyncedCondition {
+		if updateErr := r.Status().Update(ctx, nabsl); updateErr != nil {
+			logger.Error(updateErr, failedUpdateStatusError)
+			return false, updateErr
+		}
+	}
+
+	return false, nil
+}
+
+// createVeleroBSL creates a VeleroBackupStorageLocation and syncs its status with NonAdminBackupStorageLocation
+func (r *NonAdminBackupStorageLocationReconciler) createVeleroBSL(ctx context.Context, logger logr.Logger, nabsl *nacv1alpha1.NonAdminBackupStorageLocation) (bool, error) {
+	if nabsl.Status.VeleroBackupStorageLocation == nil ||
+		nabsl.Status.VeleroBackupStorageLocation.NACUUID == constant.EmptyString {
+		return false, nil
+	}
+
+	veleroObjectsNACUUID := nabsl.Status.VeleroBackupStorageLocation.NACUUID
+
+	// Check if VeleroBackupStorageLocation already exists
+	veleroBsl, err := function.GetVeleroBackupStorageLocationByLabel(ctx, r.Client, r.OADPNamespace, veleroObjectsNACUUID)
+	if err != nil {
+		logger.Error(err, "Failed to get VeleroBackupStorageLocation", constant.UUIDString, veleroObjectsNACUUID)
+		return false, err
+	}
+	// Get the VeleroBackupStorageLocation secret to be used as the credential for the VeleroBackupStorageLocation
+	veleroBslSecret, err := function.GetBslSecretByLabel(ctx, r.Client, r.OADPNamespace, veleroObjectsNACUUID)
+
+	if err != nil {
+		logger.Error(err, findSingleVBSLSecretError, constant.UUIDString, veleroObjectsNACUUID)
+		return false, err
+	}
+
+	if veleroBslSecret == nil {
+		logger.Error(err, "Failed to get VeleroBackupStorageLocation secret", constant.UUIDString, veleroObjectsNACUUID)
+		return false, err
+	}
+
+	// Create VeleroBackupStorageLocation
+	if veleroBsl == nil {
+		logger.Info("Velero BSL with label not found, creating one", "oadpnamespace", r.OADPNamespace, constant.UUIDString, veleroObjectsNACUUID)
+
+		veleroBsl = builder.ForBackupStorageLocation(r.OADPNamespace, veleroObjectsNACUUID).
+			ObjectMeta(
+				builder.WithLabels(
+					constant.NabslOriginNACUUIDLabel, veleroObjectsNACUUID,
+				),
+				builder.WithLabelsMap(function.GetNonAdminLabels()),
+				builder.WithAnnotationsMap(function.GetNonAdminBackupStorageLocationAnnotations(nabsl.ObjectMeta)),
+			).Result()
+	}
+
+	bslSpec := nabsl.Spec.DeepCopy()
+
+	// We use Credential from the secret created in the syncSecrets function
+	// however we need to set the key to the one specified in the NonAdminBackupStorageLocation spec
+	// because it's the user who decides which key to use from the secret
+	op, err := controllerutil.CreateOrUpdate(ctx, r.Client, veleroBsl, func() error {
+		veleroBsl.Spec.AccessMode = bslSpec.BackupStorageLocationSpec.AccessMode
```

Comment:
this structure makes it hard to maintain it

why not do the same way as backup and restore?
https://github.com/migtools/oadp-non-admin/blob/25e40d6abd8c49c43a1a89a9f49d6d489d07707e/internal/controller/nonadminbackup_controller.go#L620-L621

https://github.com/migtools/oadp-non-admin/blob/25e40d6abd8c49c43a1a89a9f49d6d489d07707e/internal/controller/nonadminrestore_controller.go#L344-L346

only field that needs to change in BSL spec compared to NABSL is credential, not?

Reply:
Changed. Note this is slightly different and we need to use CreateOrUpdate function here. In the Restore or Backup situation user can not modify created object as it will not be reflected on the final object state (sync controller will get the backup from the s3 for example and the user's changes are overriden).

This is NOT the case in the BSL scenario, where the user can modify settings such as credentials and they will get updated and reflected in the Velero BSL case, so we should behave similarly.